data_safe_haven/utility/logger.py

```diff
@@ -0,0 +1,234 @@
+import io
+import logging
+from typing import Any, List, Optional
+
+# Third party imports
+from rich.console import Console
+from rich.highlighter import RegexHighlighter
+from rich.logging import RichHandler
+from rich.prompt import Confirm, Prompt
+from rich.table import Table
+from rich.text import Text
+
+# Local imports
+from .types import PathType
+
+
+class LoggingHandlerPlainFile(logging.FileHandler):
+    """
+    Logging handler that cleans messages before sending them to a log file.
+    """
+
+    def __init__(
+        self, fmt: str, datefmt: str, filename: str, *args: Any, **kwargs: Any
+    ):
+        """Constructor"""
+        super().__init__(filename=filename, *args, **kwargs)
+        self.setFormatter(logging.Formatter(self.strip_formatting(fmt), datefmt))
+
+    @staticmethod
+    def strip_formatting(input_string: str) -> str:
+        """Strip console markup formatting from a string"""
+        text = Text.from_markup(input_string)
+        text.spans = []
+        return str(text)
+
+    def emit(self, record: logging.LogRecord) -> None:
+        """Emit a record without formatting"""
+        record.msg = self.strip_formatting(record.msg)
+        super().emit(record)
+
+
+class LoggingHandlerRichConsole(RichHandler):
+    """
+    Logging handler that uses Rich.
+    """
+
+    def __init__(self, fmt: str, datefmt: str, *args: Any, **kwargs: Any):
+        super().__init__(
+            *args,
+            highlighter=LogLevelHighlighter(),
+            markup=True,
+            omit_repeated_times=False,
+            rich_tracebacks=True,
+            show_level=False,
+            show_time=False,
+            tracebacks_show_locals=True,
+            **kwargs,
+        )
+        self.setFormatter(logging.Formatter(fmt, datefmt))
+
+
+class LogLevelHighlighter(RegexHighlighter):
+    """
+    Highlighter that looks for [level-name] and applies default formatting.
+    """
+
+    base_style = "logging.level."
+    highlights = [
+        r"(?P<critical>\[CRITICAL\])",
+        r"(?P<debug>\[   DEBUG\])",
+        r"(?P<error>\[   ERROR\])",
+        r"(?P<info>\[    INFO\])",
+        r"(?P<warning>\[ WARNING\])",
+    ]
+
+
+class RichStringAdaptor:
+    """
+    A wrapper to convert Rich objects into strings.
+    """
+
+    def __init__(self, coloured=False):
+        self.string_io = io.StringIO()
+        self.console = Console(file=self.string_io, force_terminal=coloured)
+
+    def to_string(self, *renderables: Any) -> str:
+        self.console.print(*renderables)
+        return self.string_io.getvalue()
+
+
+class Logger:
+    """
+    Logging singleton that can be used for anything needing logging
+    """
+
+    date_fmt = r"%Y-%m-%d %H:%M:%S"
+    rich_format = r"[log.time]%(asctime)s[/] [%(levelname)8s] %(message)s"
+    _instance: Optional["Logger"] = None
+
+    def __new__(
+        cls, verbosity: Optional[int] = None, log_file: Optional[PathType] = None
+    ) -> "Logger":
```

Comment:
I think it might be clearer just to create an instance of this class and import it where needed rather than instantiating it elsewhere in the code. Particularly because there is no `__init__` method, it is just a collection of class data and functions?

Reply:
OK, still don't feel great about writing our own singleton implementation for the sake of a few methods.

Let's not block here though.

Reply:
Sure. Feel free to open an issue for this but we'll leave it alone in this PR.